### ISSUE TITLE
--no-split-newline option for string command

### ISSUE THREAD

**Issue by mattdutson** [fish-shell/fish-shell#7227]
From the documentation of `string`:

> STRING arguments are taken from the command line unless standard input is connected to a pipe or a file, in which case they are read from standard input, one STRING per line.

Here's an example of when newline splitting is a problem. Consider `file.txt` with contents `1\n2`. If we run `cat file.txt | string replace 1\n ''`, the output is `1\n2` and not the desired `2`. This is because stdin is split on the newline and treated as two separate strings by `string replace`. The result is then re-joined with a newline when printed to stdout.

In this example, we could do `string replace 1\n '' (cat file.txt | string collect)`, but is this not unsafe for large `file.txt`? We could also do `sed -z 's/1\n//' file.txt`, but the `-z` option is not universal (macOS `sed` doesn't have it).

What I'm envisioning is a `--no-split-newline` option which would cause stdin to be treated as a single string. In the example we could do `cat file.txt | string replace -n 1\n ''` and would get the output `2`.

**Comment by krobelus** [comment]
> `string replace 1\n '' (cat file.txt | string collect)`, but isn't this unsafe for large file.txt

Builtins like `string` first read all their input and then start processing them, even when piping to them.
You can observe this by trying something like `cat | string trim`.

It is true that the command substitution will hit a size limit for large files > 128MB, but that limit is arbitrary, not imposed by the OS. Since `string` would keep everything in memory anway, I think the `string collect` approach is okay. A `--no-split-newline` would need to buffer everything as well. For large files I'd use something like `awk`.

**Comment by mattdutson** [comment]
So you're saying that since the entire contents of `file.txt` are being buffered in memory either way, a pipeline is not necessarily better here.

**Comment by mattdutson** [comment]
A couple of things to consider:
- There are cases where, even

**Comment by mattdutson** [comment]
A few things to consider:

- There are cases when a pipeline is a more natural choice and more readable than a command substitution.
- If `string` is waiting for all its input anyway, then the `--no-split-newline` option wouldn't result in additional memory overhead.
- The state of `string` currently is "you can't perform operations on newlines if input is from stdin." This seems non-ideal. Shouldn't the available functionality/power be the same regardless of the context of the invocation?

**Comment by mattdutson** [comment]
If people think this is a good idea I'd be happy to try my hand at implementing it. I've been thinking about getting involved with Fish development, and this seems simple enough to be a good starter issue.

**Comment by krobelus** [comment]
> So you're saying that since the entire contents of file.txt are being buffered in memory either way, a pipeline is not necessarily better here.

Exactly.
When operating on multiple lines at once like in your example it's hard to know how much context to keep when buffering.
Consider `string replace -r 'a\n{10}b' ''`. It's easy to just take in all.

> There are cases when a pipeline is a more natural choice and more readable than a command substitution.
> If string is waiting for all its input anyway, then the --no-split-newline option wouldn't result in additional memory overhead.

Correct.

> The state of string currently is "you can't perform operations on newlines if input is from stdin." This seems non-ideal. Shouldn't the available functionality/power be the same regardless of the context of the invocation?

I agree. The newline splitting is a convention that works really well most of the time, but definitely falls short sometimes.

The traditional Unix way to implement this would be to honor `IFS`; we have that in some places, see `read --help`, but the plan is to move away from it. Commandline options seem like a reasonable alternative. Though it would be nice if there was a consistent interface with what is now `string collect` - funnily enough piping through `string collect` does not disable splitting, only if it's last in a command substitution.

#5942 is asking for a way to disable newline splitting of output and #159 which suggests `$()` as shorthand for piping to `string collect`. (What if `$|` did the same for piped input! :imp:)

> If people think this is a good idea I'd be happy to try my hand at implementing it. I've been thinking about getting involved with Fish development, and this seems simple enough to be a good starter issue.

Probably not the most obvious issue to start with, but sure! Let's hear some opinions.

**Comment by ridiculousfish** [comment]
A point of clarification: builtins receive their input through a file descriptor, and `builtin_string` does not read it all up-front. So input is not buffered.

However builtins do buffer their full *output*, a fact I hope to fix!

**Comment by mattdutson** [comment]
> The traditional Unix way to implement this would be to honor IFS

Didn't think of this but definitely an option.

> Though it would be nice if there was a consistent interface with what is now string collect - funnily enough piping through string collect does not disable splitting, only if it's last in a command substitution.

I've gotta say I'm not a fan of the `string collect` solution to #159. It's like, "command substitutions are split on newlines, unless the last command in the pipeline is `string collect`." IMO command substitution splitting should be controlled by the command substitution syntax, not by another seemingly unrelated command. Doing `$|` seems like the same thing to me. It's the `|` operator trying to tell `string` how to split its input.

**Comment by krobelus** [comment]
Yeah `string collect` is a bit surprising. I'd ogling at the `$()` shortand for its effective bash compatibility but I hardly ever need the `string collect` behavior.

Anyway, the `--no-split-newline` behavior seems reasonable. I'd implement it as an option along the lines of
`--input-separator=`, so it works like `IFS`, just for `string`.
Of course you have to keep in mind that when using no separator, `string` will have to buffer all input unless you optimize for the specific pattern (for example your motivating pattern only needs two lines at a time), which seems too difficult.

Alternative approach: instead of the input separator, one could specify a regex which `string` then uses to repeatedly match input records. Probably too complicated and still inflexible.
